pennylane/compiler/python_compiler/xdsl_extras/constraints.py

```
@@ -16,76 +16,147 @@
+
+        * By providing an ``IRDLAttrConstraint`` for the ``rank`` argument.
+        * By providing an ``int`` representing the concrete expected rank for the ``rank``
+          argument.
+        * By providing a sequence of ``int``\ s specifying the various allowed ranks for the
+          ``rank`` argument.
 
-    Subclasses must provide 'expected_type' and 'type_name'.
+    .. note::
+
+        Either the ``shape`` or ``rank`` constraint must be provided, not both.
+
+    Args:
+        element_type (IRDLAttrConstraint | None): The constraint for the element type.
+            Default is ``None``, which indicates that any element type is allowed.
+        shape (IRDLAttrConstraint | Sequence[int] | None): The constraint for the shape.
+            Default is ``None``, which indicates that any shape is allowed.
+        shape (IRDLAttrConstraint | Sequence[int] | int | None): The constraint for the
+            rank. Default is ``None``, which indicates that any rank is allowed.
     """
 
-    expected_rank: int
-    """The expected rank."""
+    element_type: IRDLAttrConstraint[AttributeInvT]
+
+    shape: IRDLAttrConstraint[AttributeInvT]
+
+    def __init__(
+        self,
+        *,
+        element_type: IRDLAttrConstraint[AttributeInvT] | None = None,
+        shape: IRDLAttrConstraint[AttributeInvT] | Iterable[int] | None = None,
+        rank: IRDLAttrConstraint[AttributeInvT] | Iterable[int] | int | None = None,
+    ):
+        element_type = element_type or AnyAttr()
+        element_type_constr = element_type
+        shape_constr = None
+
+        if shape is not None and rank is not None:
+            raise ValueError("Only one of 'shape' or 'rank' must be provided.")
+
+        if shape is None and rank is None:
+            shape_constr = AnyAttr()
+
+        elif shape is not None:
+            shape_constr = shape
+            if isinstance(shape_constr, Iterable):
+                shape_constr = ArrayAttr([IntAttr(s) for s in shape])
```

Related to my comment above, should we instead check it `shape` is a `Collection` or `Sequence`? Also, should we raise an exception/`TypeError` if `shape` is not None and is also neither an `IRDLAttrConstraint` nor iterable?

---

We can raise a custom error, yes. I did add an assertion later in the file to check that an error is raised if, after all the processing, the shape constraint is not an instance of `Attribute | AttrConstraint` ([here](https://github.com/PennyLaneAI/pennylane/blob/3b391a61acdc9d5241ac9c25869b3b95c3001214/pennylane/compiler/python_compiler/xdsl_extras/constraints.py#L113)).

Side note, I couldn't use `isinstance(constr, IRDLAttrConstraint)` because `IRDLAttrConstraint` is a type alias containing generic aliases, which cannot be used with `isinstance`. But, these constraints are for us to use, so `Attribute | AttrConstraint` should be good enough.